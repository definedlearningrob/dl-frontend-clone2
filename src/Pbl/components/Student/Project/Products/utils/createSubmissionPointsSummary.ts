export const createSubmissionPointsSummary = (pointsAvailable: number, pointsScored: number) =>
  `${pointsScored}/${pointsAvailable}`;
