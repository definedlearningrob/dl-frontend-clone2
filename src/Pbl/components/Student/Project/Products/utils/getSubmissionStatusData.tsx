import { capitalize } from 'lodash-es';

import { PRODUCT_SUBMISSION_STATUS } from '@pbl/resources/enums';

import { SUBMISSION_STATUS } from '@shared/components/SubmissionStatusBadge';

export const getSubmissionStatusData = (
  status: PRODUCT_SUBMISSION_STATUS,
  gradePoints?: string | null
) => {
  const statusLabel = capitalize(status).replace(/[_]/g, ' ');
  const label = [statusLabel, gradePoints].filter(Boolean).join(' ');

  switch (status) {
    case PRODUCT_SUBMISSION_STATUS.DRAFT:
      return { label, value: SUBMISSION_STATUS.DRAFT };
    case PRODUCT_SUBMISSION_STATUS.SUBMITTED:
      return {
        label,
        value: SUBMISSION_STATUS.SUBMITTED,
      };
    case PRODUCT_SUBMISSION_STATUS.GRADED:
      return { label, value: SUBMISSION_STATUS.GRADED };
    case PRODUCT_SUBMISSION_STATUS.NOT_STARTED:
      return {
        label,
        value: SUBMISSION_STATUS.NOT_STARTED,
      };
  }
};
