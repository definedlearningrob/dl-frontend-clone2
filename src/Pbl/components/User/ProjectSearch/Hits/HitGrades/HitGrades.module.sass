@import '@shared/stylesheets/config.sass'

.gradesSummary
  border-radius: $border-radius-lg
  border: $border-width-xs solid $color-background
  margin: 0 $m-xs 0 0
  padding: $m-xxs $m-xs
