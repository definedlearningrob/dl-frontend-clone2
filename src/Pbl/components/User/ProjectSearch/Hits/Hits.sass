@import '@shared/stylesheets/config.sass'

.search-results
  display: grid
  gap: $m-md
  grid-template-columns: repeat(auto-fill, minmax(240px, 1fr))
  margin: 0
  width: 100%
