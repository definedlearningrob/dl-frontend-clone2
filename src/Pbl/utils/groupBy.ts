type TItemsContains = {
  items: any[];
};

export function groupBy<T extends TItemsContains>(list: object[], groupBy: string[]): T[] {
  const output = list.reduce((acc: T[], curr: any) => {
    const foundIndex = acc.findIndex((group: any) =>
      groupBy.every((byItem) => group[byItem] === curr[byItem])
    );
    const elementNotFound = foundIndex === -1;
    if (elementNotFound) {
      const newObject = groupBy.reduce((acc, next) => ({ ...acc, [next]: curr[next] }), {});
      acc.push({ ...newObject, items: [curr] } as any as T);
    } else {
      acc[foundIndex].items.push(curr);
    }

    return acc;
  }, []);

  return output;
}
