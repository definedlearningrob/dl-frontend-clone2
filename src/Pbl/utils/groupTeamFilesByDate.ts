import dayjs from 'dayjs';

import { SubmissionContent } from '@shared/components/Submissions/helpers/types';
import { TFile } from '@shared/resources/types';

type SubmissionContentWithFiles = Omit<SubmissionContent, 'content'> & {
  content: TFile[];
};

const MAX_TIME_DIFF_TO_GROUP = 300; // 5 minutes in seconds

export const groupFilesByDate = (files: TFile[]) => {
  const groupedFiles = files.reduce((acc, file, currIndex) => {
    const prevItem = acc[currIndex - 1] || acc.at(-1);

    // last item is from same person
    if (prevItem?.author.uuid === file.submitter.uuid) {
      const currentTime = file.createdAt;
      const prevTime = prevItem.date;

      const diff = dayjs(currentTime).diff(prevTime, 'second');

      const isWithinTime = Math.abs(diff) < MAX_TIME_DIFF_TO_GROUP;

      if (isWithinTime && prevItem.content) {
        prevItem.content.push(file);
        prevItem.date = file.createdAt;

        return acc;
      }
    }

    acc.push({
      id: file.id,
      author: {
        name: `${file.submitter.firstName} ${file.submitter.lastName}`,
        uuid: file.submitter.uuid,
      },
      date: file.createdAt,
      content: [file],
    });

    return acc;
  }, [] as any as SubmissionContentWithFiles[]);

  return groupedFiles;
};
