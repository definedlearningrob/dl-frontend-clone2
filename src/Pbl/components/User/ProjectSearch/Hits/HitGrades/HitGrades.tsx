import { useMemo } from 'react';

import { sortGrades } from '@pbl/components/User/ProjectSearch/helpers';
import { TGrade } from '@pbl/graphql/user/queries/dashboardCourses';
import { GradesListElement } from '@pbl/components/User/Dashboard/Courses/Card/GradesListElement';

import { groupGradesByRange } from '@shared/utils/groupGradesByRange';

import styles from './HitGrades.module.sass';

type Props = {
  grades: TGrade[];
};

export const HitGrades = ({ grades }: Props) => {
  const sortedGrades = sortGrades(grades);
  const groupedGradesByRange = useMemo(() => groupGradesByRange(sortedGrades), [grades]);

  const renderGrades = () =>
    groupedGradesByRange.map((range) => <GradesListElement key={range.start} range={range} />);

  return (
    <ul className={styles.gradesSummary} data-testid='grades-summary'>
      {renderGrades()}
    </ul>
  );
};
