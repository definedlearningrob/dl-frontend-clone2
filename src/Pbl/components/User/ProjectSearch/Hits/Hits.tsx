import { connectInfiniteHits, Hit, InfiniteHitsProvided } from 'react-instantsearch-core';
import { useEffect, useMemo, useRef, useState } from 'react';

import HitComponent from './Hit/Hit';
import './Hits.sass';

const UserProjectSearchHits = ({ hits, hasMore, refineNext }: InfiniteHitsProvided) => {
  const sentinel = useRef<HTMLDivElement | null>(null);
  const [isIntersecting, setIsIntersectiing] = useState(false);

  const onSentinelIntersection = (entries: IntersectionObserverEntry[]) => {
    entries.forEach((entry: IntersectionObserverEntry) => {
      setIsIntersectiing(entry.isIntersecting);
    });
  };

  const observer = useMemo(() => new IntersectionObserver(onSentinelIntersection), []);

  useEffect(() => {
    if (sentinel.current) {
      observer.observe(sentinel.current);
    }

    () => observer.disconnect();
  }, [sentinel.current, hits]);

  useEffect(() => {
    if (hasMore && isIntersecting) {
      refineNext();
    }
  }, [isIntersecting]);

  return (
    <div className='search-results'>
      {hits.map((hit: Hit) => (
        <HitComponent key={hit.objectID} hit={hit} />
      ))}
      <div ref={sentinel} />
    </div>
  );
};

export default connectInfiniteHits(UserProjectSearchHits);
