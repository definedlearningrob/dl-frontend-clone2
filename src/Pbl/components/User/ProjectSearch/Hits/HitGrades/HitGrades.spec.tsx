import { renderWithRouter } from '@pbl/utils/test';
import { HitGrades } from '@pbl/components/User/ProjectSearch/Hits/HitGrades';

describe('HitGrades', () => {
  const mockedGrades = ['K', '1', '2', '3', '4', '5'];

  it('should display grades', () => {
    // @ts-ignore
    const { getByTestId, getByText } = renderWithRouter(<HitGrades grades={mockedGrades} />);

    expect(getByTestId('grades-summary')).toBeInTheDocument();
    expect(getByText('Grades: K-5')).toBeInTheDocument();
  });

  it('should render correctly', () => {
    // @ts-ignore
    const { container } = renderWithRouter(<HitGrades grades={mockedGrades} />);

    expect(container).toMatchSnapshot();
  });
});
