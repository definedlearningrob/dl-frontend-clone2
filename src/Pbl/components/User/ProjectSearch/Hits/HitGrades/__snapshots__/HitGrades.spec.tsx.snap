// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HitGrades should render correctly 1`] = `
<div>
  <ul
    class="gradesSummary"
    data-testid="grades-summary"
  >
    <li
      class="grade"
      data-testid="grade-list-element"
    >
      <p
        class="gradeText"
        data-testid="grade-text"
      >
        Grades: K-5
      </p>
    </li>
  </ul>
</div>
`;
